src/libraries/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -1053,5 +1079,62 @@ internal void ProcessBytes(int numBytes)
                 Buffer = Buffer.Slice(numBytes);
             }
         }
+
+        // Use a custom queue instead of ConcurrentQueue because ConcurrentQueue preserves segments for
+        // observation when using TryPeek(). These segments will not clear out references after a dequeue
+        // and as a result they hold on to SerialStreamIORequest instances so that they cannot be GC'ed.
+        // This in turn means that any buffers that the client supplied are not eligible for GC either.
```

Comment:
honestly this sounds more like a bug with ConcurrentQueue implementation, could you file an issue about this? If we get confirmation this is a bug then we can temporarily use workaround until issue is fixed

Reply:
also I'm ok with using queue and locks directly in the code because at times it makes sense to put lock around larger piece of code than on each operation individually (i.e. RemoveCompletedTasks could be locked entirely)

Reply:
it was discussed in https://github.com/dotnet/runtime/issues/28953

Reply:
I'd still consider using locks directly though as I think that can be done much more efficiently than lock per operation

Reply:
I can have a look at that this week and update the PR :)

Reply:
Finally updated this PR, now locks are used directly in the code